### ISSUE TITLE
Workshop mods fail to load

### ISSUE THREAD

**Issue by Pingperfect** [Tyrsis/SE-Community-Mod-API#29]
Apparently Github does not like long messages so had to post this on our forums to properly explain. Please see issue here http://pingperfect.com/phpbb/viewtopic.php?f=64&t=646

**Comment by Tyrsis** [comment]
Extender has 0 control over mod download.  It's all done through steam. I've seen this issue before and it's steam related (steam was down, or intermittent availability).  Ask them to try again later and if it still happens, then I'm nto entirely sure what to tell you, as there is 0 reference for me to try to fix this as at that point in loading extender has done nothing except launch the game.

**Comment by Pingperfect** [comment]
Wow fast reply :)

Those logs are minutes apart, no issues with steam and this is a consistent issue across all our boxes in several different countries. It's defiantly somthing that sese is doing/changing but as to what, I'm not sure

**Comment by Tyrsis** [comment]
I see the logs, the issue is I've had it where it wouldn't work 4 times out of 5, but then the 5th time it worked.  Like I said, there is literally nothing I can do about this problem as at this point in extender, all it's done is run the dedicated server itself.

**Comment by Tyrsis** [comment]
The only thing I can think of is for some reason these are loading two different .cfg files.  The .cfg is where the mods are.  (well both the sandbox and the .cfg)

**Comment by McyD** [comment]
I have been having issues with this for the last 2-3 weeks.  Server shows connection to steam fine, but mods fail to download.  If nothing can be done, can a check be added that if mods exist they do not fail, if so force another restart.

Loading a server where the mods fail, mean that the server is unreachable and no one can join, even if it is running.  This currently results in an automated restart causing a server to load to which no one can access.

**Comment by dodexahedron** [comment]
Can you provide an example Sandbox.sbc and SpaceEngineers-Dedicated.cfg for a world that has this problem?  I've never been able to reproduce this behavior, and we've got ~40 steam workshop mods on our server.

**Comment by McyD** [comment]
https://www.dropbox.com/sh/rdetrzt3n5lptj2/AAAN5cb2t03LgaYzy1rYhK7_a?dl=0

**Comment by Tyrsis** [comment]
I would agree McyD, the problem is, it's the dedicated server itself who decides to just give up, not extender.  I'm not sure I can work around that limitation.  But I can certain take a look.

**Comment by McyD** [comment]
I don't know if you can capture it or not, but the log does show "Mods failed to download".  Thought it might be possible to detect that and force a restart if it is found during the last start-up.

**Comment by dodexahedron** [comment]
Here's an idea that would help this:
Automatic workshop content updating.

It would work like this:
On a scheduled interval, the server checks for updated workshop mods from the list of mods it is already aware of.  If any are in need of updating, the server could do a graceful save and restart.
This could be customized by allowing the admin to specify which mods the updating behavior is allowed for (maybe you don't care if the skybox is out of date, but you REALLY want that arc reactor pack to keep working or whatever).
Allowing the admin to specify the update schedule would keep it from being (too much of) a nuisance to users, and would ensure that missed mod updates get pulled in on a schedule that admins consider to be acceptable.

**Comment by McyD** [comment]
Hmm, doesn't the client always download the latest version of the mods from steam when connecting?  Or does the client downgrade the mod version to match the servers version in-case of a newer version on steam?  Guessing matches the server since I have never had an issue that a client can't connect due to an out of date mod.

I have a restart set for every 6 hours so this would not help me anyway, though I can see it being useful to others.  Really, this should be a separate issue I think.  Especially, since the mod failed to download issue with steam would still effect it an auto-updating mod system.

Just to clarify the "mod failed to download" issue, is not an mod updating issue per say. It just means the connection to the steam workshop failed, even though the connection to steam was successful, and the server cannot even check if the mods are up to date or not.  When this happens the mods on the server are not loaded, the server is not listed in the servers list, and no one can join; even though the map successfully loads and the server is up.

**Comment by dodexahedron** [comment]
Right.  The point being that an "auto-updater" would not bother with needless restarts of the server until steam is working properly.

**Comment by McyD** [comment]
Granted, but it would not fix the issue of the workshop failing during normal restarts... which was the point on this issue.

If the server cannot check the workshop during startup, though it is connected to steam, then the server just posts "failed to download mods" and continues its normal startup... as I have stated before, thus the issue for a check to be implemented.
